### ISSUE TITLE
table_name_prefix and table_name_suffix changes column names which cause ORA-00904

### ISSUE THREAD

**Issue by yahonda** [rsim/oracle-enhanced#639]
Tested Oracle enhanced adapter master branch with Rails 4-2-stable, these two test get ORA-904 error.

```sql
     ActiveRecord::StatementInvalid:
       OCIError: ORA-00904: "XXX_TEST_POST_ID": invalid identifier: ALTER TABLE "XXX_TEST_COMMENTS" ADD CONSTRAINT "FK_RAILS_1425FC0AB5"
       FOREIGN KEY ("XXX_TEST_POST_ID")
         REFERENCES "XXX_TEST_POSTS" ("ID")
```

This SQL causes ORA-904 since `XXX_TEST_POST_ID` does not exist at `XXX_TEST_COMMENTS` table.

```sql
ALTER TABLE "XXX_TEST_COMMENTS" ADD CONSTRAINT "FK_RAILS_1425FC0AB5"
       FOREIGN KEY ("XXX_TEST_POST_ID")
         REFERENCES "XXX_TEST_POSTS" ("ID")
```
It should be as follows. column name at `XXX_TEST_COMMENTS` table is `TEST_POST_ID`.

```sql
ALTER TABLE "XXX_TEST_COMMENTS" ADD CONSTRAINT "FK_RAILS_1425FC0AB5"
       FOREIGN KEY ("TEST_POST_ID")
         REFERENCES "XXX_TEST_POSTS" ("ID")
```

Here column name `XXX_TEST_POST_ID` is generated at `foreign_key_column_for` https://github.com/rails/rails/blob/4-2-stable/activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb#L804-L806

```ruby
      def foreign_key_column_for(table_name) # :nodoc:
        "#{table_name.to_s.singularize}_id"
      end
```

This name is based on `table_name` arguments, which is affected by `table_name_prefix` and `table_name_suffix`. Based on the naming, I do not think `table_name_prefix` and `table_name_suffix` should not change column names.

```ruby
$ rspec ./spec/active_record/connection_adapters/oracle_enhanced_schema_statements_spec.rb:643

==> Loading config from ENV or use default
==> Running specs with MRI version 2.2.2
==> Selected Rails version 4.0-master
==> Effective ActiveRecord version 4.2.2
Run options: include {:locations=>{"./spec/active_record/connection_adapters/oracle_enhanced_schema_statements_spec.rb"=>[643]}}
F

Failures:

  1) OracleEnhancedAdapter schema definition foreign key constraints with table_name_prefix should use table_name_prefix for foreign table
     Failure/Error: add_foreign_key :test_comments, :test_posts
     ActiveRecord::StatementInvalid:
       OCIError: ORA-00904: "XXX_TEST_POST_ID": invalid identifier: ALTER TABLE "XXX_TEST_COMMENTS" ADD CONSTRAINT "FK_RAILS_1425FC0AB5"
       FOREIGN KEY ("XXX_TEST_POST_ID")
         REFERENCES "XXX_TEST_POSTS" ("ID")
     # stmt.c:243:in oci8lib_220.so
     # /home/yahonda/git/ruby-oci8/lib/oci8/cursor.rb:129:in `exec'
     # /home/yahonda/git/ruby-oci8/lib/oci8/oci8.rb:278:in `exec_internal'
     # /home/yahonda/git/ruby-oci8/lib/oci8/oci8.rb:269:in `exec'
     # ./lib/active_record/connection_adapters/oracle_enhanced/oci_connection.rb:431:in `exec'
     # ./lib/active_record/connection_adapters/oracle_enhanced/oci_connection.rb:90:in `exec'
     # ./lib/active_record/connection_adapters/oracle_enhanced/database_statements.rb:11:in `block in execute'
     # /home/yahonda/git/rails/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:473:in `block in log'
     # /home/yahonda/git/rails/activesupport/lib/active_support/notifications/instrumenter.rb:20:in `instrument'
     # /home/yahonda/git/rails/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:467:in `log'
     # ./lib/active_record/connection_adapters/oracle_enhanced_adapter.rb:1348:in `log'
     # ./lib/active_record/connection_adapters/oracle_enhanced/database_statements.rb:11:in `execute'
     # /home/yahonda/git/rails/activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb:762:in `add_foreign_key'
     # ./lib/active_record/connection_adapters/oracle_enhanced/schema_statements.rb:411:in `add_foreign_key'
     # /home/yahonda/git/rails/activerecord/lib/active_record/migration.rb:661:in `block in method_missing'
     # /home/yahonda/git/rails/activerecord/lib/active_record/migration.rb:631:in `block in say_with_time'
     # /home/yahonda/.rvm/rubies/ruby-2.2.2/lib/ruby/2.2.0/benchmark.rb:288:in `measure'
     # /home/yahonda/git/rails/activerecord/lib/active_record/migration.rb:631:in `say_with_time'
     # /home/yahonda/git/rails/activerecord/lib/active_record/migration.rb:651:in `method_missing'
     # ./spec/active_record/connection_adapters/oracle_enhanced_schema_statements_spec.rb:645:in `block (5 levels) in <top (required)>'
     # ./spec/spec_helper.rb:117:in `instance_eval'
     # ./spec/spec_helper.rb:117:in `block (2 levels) in schema_define'
     # /home/yahonda/git/rails/activerecord/lib/active_record/migration.rb:639:in `suppress_messages'
     # ./spec/spec_helper.rb:116:in `block in schema_define'
     # /home/yahonda/git/rails/activerecord/lib/active_record/schema.rb:41:in `instance_eval'
     # /home/yahonda/git/rails/activerecord/lib/active_record/schema.rb:41:in `define'
     # /home/yahonda/git/rails/activerecord/lib/active_record/schema.rb:61:in `define'
     # ./spec/spec_helper.rb:115:in `schema_define'
     # ./spec/active_record/connection_adapters/oracle_enhanced_schema_statements_spec.rb:644:in `block (4 levels) in <top (required)>'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/extensions/instance_eval_with_args.rb:16:in `instance_exec'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/extensions/instance_eval_with_args.rb:16:in `instance_eval_with_args'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/example.rb:116:in `block in run'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/example.rb:248:in `with_around_each_hooks'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/example.rb:113:in `run'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/example_group.rb:515:in `block in run_examples'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/example_group.rb:511:in `map'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/example_group.rb:511:in `run_examples'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/example_group.rb:496:in `run'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/example_group.rb:497:in `block in run'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/example_group.rb:497:in `map'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/example_group.rb:497:in `run'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/example_group.rb:497:in `block in run'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/example_group.rb:497:in `map'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/example_group.rb:497:in `run'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/command_line.rb:24:in `block (2 levels) in run'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/command_line.rb:24:in `map'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/command_line.rb:24:in `block in run'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/reporter.rb:58:in `report'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/command_line.rb:21:in `run'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/runner.rb:103:in `run'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/runner.rb:17:in `block in autorun'

Finished in 0.23778 seconds
1 example, 1 failure

Failed examples:

rspec ./spec/active_record/connection_adapters/oracle_enhanced_schema_statements_spec.rb:643 # OracleEnhancedAdapter schema definition foreign key constraints with table_name_prefix should use table_name_prefix for foreign table
$
```

```ruby
$ rspec ./spec/active_record/connection_adapters/oracle_enhanced_schema_statements_spec.rb:657

==> Loading config from ENV or use default
==> Running specs with MRI version 2.2.2
==> Selected Rails version 4.0-master
==> Effective ActiveRecord version 4.2.2
Run options: include {:locations=>{"./spec/active_record/connection_adapters/oracle_enhanced_schema_statements_spec.rb"=>[657]}}
F

Failures:

  1) OracleEnhancedAdapter schema definition foreign key constraints with table_name_suffix should use table_name_suffix for foreign table
     Failure/Error: add_foreign_key :test_comments, :test_posts
     ActiveRecord::StatementInvalid:
       OCIError: ORA-00904: "TEST_POSTS_XXX_ID": invalid identifier: ALTER TABLE "TEST_COMMENTS_XXX" ADD CONSTRAINT "FK_RAILS_89980630D2"
       FOREIGN KEY ("TEST_POSTS_XXX_ID")
         REFERENCES "TEST_POSTS_XXX" ("ID")
     # stmt.c:243:in oci8lib_220.so
     # /home/yahonda/git/ruby-oci8/lib/oci8/cursor.rb:129:in `exec'
     # /home/yahonda/git/ruby-oci8/lib/oci8/oci8.rb:278:in `exec_internal'
     # /home/yahonda/git/ruby-oci8/lib/oci8/oci8.rb:269:in `exec'
     # ./lib/active_record/connection_adapters/oracle_enhanced/oci_connection.rb:431:in `exec'
     # ./lib/active_record/connection_adapters/oracle_enhanced/oci_connection.rb:90:in `exec'
     # ./lib/active_record/connection_adapters/oracle_enhanced/database_statements.rb:11:in `block in execute'
     # /home/yahonda/git/rails/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:473:in `block in log'
     # /home/yahonda/git/rails/activesupport/lib/active_support/notifications/instrumenter.rb:20:in `instrument'
     # /home/yahonda/git/rails/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:467:in `log'
     # ./lib/active_record/connection_adapters/oracle_enhanced_adapter.rb:1348:in `log'
     # ./lib/active_record/connection_adapters/oracle_enhanced/database_statements.rb:11:in `execute'
     # /home/yahonda/git/rails/activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb:762:in `add_foreign_key'
     # ./lib/active_record/connection_adapters/oracle_enhanced/schema_statements.rb:411:in `add_foreign_key'
     # /home/yahonda/git/rails/activerecord/lib/active_record/migration.rb:661:in `block in method_missing'
     # /home/yahonda/git/rails/activerecord/lib/active_record/migration.rb:631:in `block in say_with_time'
     # /home/yahonda/.rvm/rubies/ruby-2.2.2/lib/ruby/2.2.0/benchmark.rb:288:in `measure'
     # /home/yahonda/git/rails/activerecord/lib/active_record/migration.rb:631:in `say_with_time'
     # /home/yahonda/git/rails/activerecord/lib/active_record/migration.rb:651:in `method_missing'
     # ./spec/active_record/connection_adapters/oracle_enhanced_schema_statements_spec.rb:659:in `block (5 levels) in <top (required)>'
     # ./spec/spec_helper.rb:117:in `instance_eval'
     # ./spec/spec_helper.rb:117:in `block (2 levels) in schema_define'
     # /home/yahonda/git/rails/activerecord/lib/active_record/migration.rb:639:in `suppress_messages'
     # ./spec/spec_helper.rb:116:in `block in schema_define'
     # /home/yahonda/git/rails/activerecord/lib/active_record/schema.rb:41:in `instance_eval'
     # /home/yahonda/git/rails/activerecord/lib/active_record/schema.rb:41:in `define'
     # /home/yahonda/git/rails/activerecord/lib/active_record/schema.rb:61:in `define'
     # ./spec/spec_helper.rb:115:in `schema_define'
     # ./spec/active_record/connection_adapters/oracle_enhanced_schema_statements_spec.rb:658:in `block (4 levels) in <top (required)>'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/extensions/instance_eval_with_args.rb:16:in `instance_exec'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/extensions/instance_eval_with_args.rb:16:in `instance_eval_with_args'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/example.rb:116:in `block in run'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/example.rb:248:in `with_around_each_hooks'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/example.rb:113:in `run'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/example_group.rb:515:in `block in run_examples'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/example_group.rb:511:in `map'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/example_group.rb:511:in `run_examples'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/example_group.rb:496:in `run'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/example_group.rb:497:in `block in run'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/example_group.rb:497:in `map'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/example_group.rb:497:in `run'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/example_group.rb:497:in `block in run'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/example_group.rb:497:in `map'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/example_group.rb:497:in `run'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/command_line.rb:24:in `block (2 levels) in run'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/command_line.rb:24:in `map'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/command_line.rb:24:in `block in run'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/reporter.rb:58:in `report'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/command_line.rb:21:in `run'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/runner.rb:103:in `run'
     # /home/yahonda/.rvm/gems/ruby-2.2.2@rails42/gems/rspec-core-2.99.2/lib/rspec/core/runner.rb:17:in `block in autorun'

Finished in 0.24933 seconds
1 example, 1 failure

Failed examples:

rspec ./spec/active_record/connection_adapters/oracle_enhanced_schema_statements_spec.rb:657 # OracleEnhancedAdapter schema definition foreign key constraints with table_name_suffix should use table_name_suffix for foreign table
$
```

**Comment by yahonda** [comment]
Opened https://github.com/rails/rails/issues/20689 since I think it needs to be fixed (or at least needs to be confirmed by Rails developers).

**Comment by yahonda** [comment]
Closed since https://github.com/rails/rails/pull/20699 has been rails master and back ported into 4-2-stable. I think this fix will be available next version of Rails 4.2.4.
